product.go

```diff
@@ -72,12 +78,99 @@ type ProductsResource struct {
 	Products []Product `json:"products"`
 }
 
+// Pagination of results
+type Pagination struct {
+	NextPageOptions		*ListOptions
+	PreviousPageOptions *ListOptions
+}
+
 // List products
 func (s *ProductServiceOp) List(options interface{}) ([]Product, error) {
+	products, _, err := s.ListWithPagination(options)
+	return products, err
+}
+
+// List products and return pagination to retrieve next/previous results
+func (s *ProductServiceOp) ListWithPagination(options interface{}) ([]Product, Pagination, error) {
 	path := fmt.Sprintf("%s/%s.json", globalApiPathPrefix, productsBasePath)
 	resource := new(ProductsResource)
-	err := s.client.Get(path, resource, options)
-	return resource.Products, err
+	headers := http.Header{}
+
+	err := s.client.CreateAndDo("GET", path, nil, options, resource, &headers)
+
+	// Extract pagination from header
+	linkHeader := ""
+	_, ok := headers["Link"]
+	if ok {
+		linkHeader = headers["Link"][0]
+	}
+	pagination, err := extractPagination(linkHeader)
+
+	return resource.Products, pagination, err
+}
+
+// Extract pagination from the Link header 
+// Details on the format are here:
+// https://help.shopify.com/en/api/guides/paginated-rest-results
+func extractPagination(linkHeader string) (Pagination, error) {
+	pagination := Pagination{}
+	
+	if linkHeader == "" {
+		return pagination, nil
+	}
+
+	linkRegex := regexp.MustCompile(`^ *<([^>]+)>; rel="(previous|next)" *$`)
+
+	for _, link := range strings.Split(linkHeader, ",") {
+		match := linkRegex.FindStringSubmatch(link); 
+		if len(match) == 0 {
```

Comment:
Should check that the `len(match) == 2` since you are using `match[1]` and `match[2]` later on. With current code I believe you could receive one hit and end up panicing when it tries to access `match[2]`.

Reply:
If a match is found then the `match` will always have 3 values, `match[0]` is the full match followed by the subexpressions. `match[1]` is the URL and `match[2]` is either 'previous' or 'next'. 
If will instead check for `nil` to make it a bit less confusing:
`if match == nil {`

Reply:
That is not guaranteed by `FindStringSubmatch`. If you pass an empty string or other bad data then this will panic.

Reply:
The documentation states a return value of nil indicates no match.
https://golang.org/pkg/regexp/#Regexp.FindStringSubmatch

Testing this it looks like an empty array is returned with no matches:
https://play.golang.org/p/7d3tXFPiyq4